python/ray/data/_internal/execution/operators/zip_operator.py

```diff
@@ -0,0 +1,238 @@
+        # Check that both sides have the same number of rows.
+        # TODO(Clark): Support different number of rows via user-directed
+        # dropping/padding.
+        total_left_rows = sum(left_block_rows)
+        total_right_rows = sum(right_block_rows)
+        if total_left_rows != total_right_rows:
+            raise ValueError(
+                "Cannot zip datasets of different number of rows: "
+                f"{total_left_rows}, {total_right_rows}"
+            )
+
+        # Whether the left and right input sides are inverted
+        input_side_inverted = False
+        if sum(right_block_bytes) > sum(left_block_bytes):
+            # Make sure that right side is smaller, so we minimize splitting
+            # work when aligning both sides.
+            # TODO(Clark): Improve this heuristic for minimizing splitting work,
+            # e.g. by generating the splitting plans for each route (via
+            # _generate_per_block_split_indices) and choosing the plan that splits
+            # the least cumulative bytes.
+            left_blocks_with_metadata, right_blocks_with_metadata = (
+                right_blocks_with_metadata,
+                left_blocks_with_metadata,
+            )
+            left_block_rows, right_block_rows = right_block_rows, left_block_rows
+            input_side_inverted = True
+
+        # Get the split indices that will align both sides.
+        indices = list(itertools.accumulate(left_block_rows))
+        indices.pop(-1)
+
+        # Split other at the alignment indices, such that for every block from
+        # left side, we have a list of blocks from right side that have the same
+        # cumulative number of rows as that left block.
+        # NOTE: _split_at_indices has a no-op fastpath if the blocks are already
+        # aligned.
+        aligned_right_blocks_with_metadata = _split_at_indices(
+            right_blocks_with_metadata,
+            indices,
+            block_rows=right_block_rows,
+        )
+        del right_blocks_with_metadata
+
+        left_blocks = [b for b, _ in left_blocks_with_metadata]
+        right_blocks_list = aligned_right_blocks_with_metadata[0]
+        del left_blocks_with_metadata, aligned_right_blocks_with_metadata
+
+        zip_one_block = cached_remote_fn(_zip_one_block, num_returns=2)
+
+        output_blocks = []
+        output_metadata = []
+        for left_block, right_blocks in zip(left_blocks, right_blocks_list):
+            # For each block from left side, zip it together with 1 or more blocks from
+            # right side. We're guaranteed to have that left_block has the same number
+            # of rows as right_blocks has cumulatively.
+            res, meta = zip_one_block.remote(
+                left_block, *right_blocks, inverted=input_side_inverted
+            )
+            output_blocks.append(res)
+            output_metadata.append(meta)
+
+        # Early release memory.
+        del left_blocks, right_blocks_list
```

Comment:
When should we call bundle.destroy_if_owned() for the input bundles?

Reply:
Yes it should be safe to call it after the output blocks materialized in object store. Added below.